enclave-runtime/src/lib.rs

```diff
@@ -724,76 +798,63 @@ fn load_initialized_state(shard: &H256) -> SgxResult<State> {
 	Ok(state)
 }
 
-fn execute_trusted_getters<P>(top_pool: &P, shards: &[H256]) -> Result<()>
+/// Execute pending trusted getters for the `shard` until `max_exec_duration` is reached.
+fn exec_trusted_getters<P>(top_pool: &P, shard: H256, max_exec_duration: Duration) -> Result<()>
 where
 	P: TrustedOperationPool<Hash = H256> + 'static,
 {
 	let author = Author::new(Arc::new(top_pool));
+	let ends_at = duration_now() + max_exec_duration;
+
+	// retrieve trusted operations from pool
+	let trusted_getters = author.get_pending_tops_separated(shard)?.1;
+	for trusted_getter_signed in trusted_getters.into_iter() {
+		let hash_of_getter = author.hash_of(&trusted_getter_signed.clone().into());
+
+		// get state
+		match get_stf_state(trusted_getter_signed, shard) {
+			Ok(r) => {
+				trace!("Successfully loaded stf state");
+				// let client know of current state
+				trace!("Updating client");
+				match top_pool.rpc_send_state(hash_of_getter, r.encode()) {
+					Ok(_) => trace!("Successfully updated client"),
+					Err(e) => error!("Could not send state to client {:?}", e),
+				}
+			},
+			Err(e) => {
+				error!("failed to get stf state, skipping trusted getter ({:?})", e);
+			},
+		};
 
-	let start_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() as i64;
-	let mut is_done = false;
-	for shard in shards.to_owned().into_iter() {
-		// retrieve trusted operations from pool
-		let trusted_getters = author.get_pending_tops_separated(shard)?.1;
-		for trusted_getter_signed in trusted_getters.into_iter() {
-			// get hash
-			let hash_of_getter = author.hash_of(&trusted_getter_signed.clone().into());
-
-			// get state
-			match get_stf_state(trusted_getter_signed, shard) {
-				Ok(r) => {
-					trace!("Successfully loaded stf state");
-					// let client know of current state
-					trace!("Updating client");
-					match top_pool.rpc_send_state(hash_of_getter, r.encode()) {
-						Ok(_) => trace!("Successfully updated client"),
-						Err(e) => error!("Could not send state to client {:?}", e),
-					}
-				},
-				Err(e) => {
-					error!("failed to get stf state, skipping trusted getter ({:?})", e);
-				},
-			};
-
-			// remove getter from pool
-			if let Err(e) =
-				author.remove_top(vec![TrustedOperationOrHash::Hash(hash_of_getter)], shard, false)
-			{
-				error!("Error removing trusted operation from top pool: Error: {:?}", e);
-			}
-			// Check time
-			if time_is_overdue(Timeout::Getter, start_time) {
-				is_done = true;
-				break
-			}
+		// remove getter from pool
+		if let Err(e) =
+			author.remove_top(vec![TrustedOperationOrHash::Hash(hash_of_getter)], shard, false)
+		{
+			error!("Error removing trusted operation from top pool: Error: {:?}", e);
 		}
-		if is_done {
-			break
+		// Check time
+		if ends_at < duration_now() {
+			return Ok(())
```

Comment:
I'm guessing you have a test for this condition now?

Reply:
Oh, your  eye sees everything. :) To be honest, I don't have any tests here. But this method is copy-pasted from my other branch with `Aura`, and I have quite some tests there. The tests will arrive with the next merge. Does this suffice for you? 😸

Reply:
I think this whole time-slotting is quite critical and easy to get wrong. If you have tests for this in the next PR, that's fine for me 👍

Reply:
You are absolutely correct. Some of the enclave tests have already uncovered previous flaws in my design.